Create "EVM vs PolkaVM" page in sc basics section (#376)
polkadot-developers/polkadot-docs: polkadot-protocol/smart-contracts-basics/evm-vs-polkavm.md

## Core Virtual Machine Architecture

The most significant departure from Ethereum comes from PolkaVM's foundation itself. Rather than implementing the EVM, PolkaVM utilizes a RISC-V instruction set. For most Solidity developers, this architectural change remains transparent thanks to the [Revive compiler's](https://github.com/paritytech/revive){target=\_blank} complete Solidity support, including inline assembler functionality.

However, this architectural difference becomes relevant in specific scenarios. Tools that attempt to download and inspect contract bytecode will fail, as they expect EVM bytecode rather than PolkaVM's RISC-V format. This primarily affects contracts using [`EXTCODECOPY`](https://www.evm.codes/?fork=cancun#3c){target=\_blank} to manipulate code at runtime, though such cases are rare. PolkaVM offers an elegant alternative through its [on-chain constructors](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/pallet/struct.Pallet.html#method.bare_instantiate){target=\_blank}, enabling contract instantiation without runtime code modification.

## Gas Model

> Ethereum's resource model relies on a single metric: [`gas`](https://ethereum.org/en/developers/docs/gas/#what-is-gas){target=\_blank}, which serves as the universal unit for measuring computational costs. Each operation on the network consumes a specific amount of gas. Most platforms aiming for Ethereum compatibility typically adopt identical gas values to ensure seamless integration.

Reviewer: ```suggestion
Ethereum's resource model relies on a single metric: [gas](https://ethereum.org/en/developers/docs/gas/#what-is-gas){target=\_blank}, which serves as the universal unit for measuring computational costs. Each operation on the network consumes a specific amount of gas. Most platforms aiming for Ethereum compatibility typically adopt identical gas values to ensure seamless integration.
```

> PolkaVM introduces two significant changes to Ethereum's gas model:

Reviewer: I'm not a fan of these sentences that end with colons and lead directly into subsections. Either briefly mention dynamic gas value scaling and multi-dimensional... or say something generic like the significant changes to Ethereum's gas model will be outlined in the following sections.

Reviewer: This is still a thing, can you please double-check it?

Author: oh I thought I addressed it here https://github.com/polkadot-developers/polkadot-docs/pull/376/commits/9746de4e972bf0b9c35bf93148679c653b12bd81#diff-5d27d7457e321fa450304dabe080cc5d332ab55bd8b4ee5ff95acd0aaa88ad2dR18

Could you please check? Thanks!

Author: sorry, actually it should be fixed here [a878192](https://github.com/polkadot-developers/polkadot-docs/pull/376/commits/a8781922c026f2a650d1a5baca249409fb58b10e)

What do you think?

Reviewer: perfect, thanks!

### Dynamic Gas Value Scaling

> Instead of adhering to Ethereum's fixed gas values, PolkaVM implements benchmark-based pricing that better reflects its improved execution performance. This makes instructions cheaper relative to I/O-bound operations but requires developers to avoid hardcoding gas values, particularly in cross-contract calls

Reviewer: ```suggestion
Instead of adhering to Ethereum's fixed gas values, PolkaVM implements benchmark-based pricing that better reflects its improved execution performance. This makes instructions cheaper relative to I/O-bound operations but requires developers to avoid hardcoding gas values, particularly in cross-contract calls.
```

- **`ref_time`** - equivalent to traditional gas, measuring computation time
- **`proof_size`** - tracks state proof size for validator verification
- **`storage_deposit`** - manages state bloat through a deposit system

These three can be limited at the transaction level, just like gas on Ethereum. The Ethereum RPC proxy maps all three of these into the single dimension gas so that everything behaves as on Ethereum for users. This ensures that the transaction cost displayed in the wallet accurately represents the actual costs, even though it uses these three resources internally.

These resources can also be limited when making a cross-contract call. However, Solidity doesn't allow specifying anything other than `gas_limit` for a cross-contract call. The PolkaVM takes the `gas_limit` the contract supplies and uses that as `ref_time_limit.` The other two resources are just uncapped in this case. Please note that uncapped means the transaction-specified limits still constrain them, so this cannot be used to trick the signer of the transaction.

Resource limiting in cross-contract calls serves a critical security purpose, particularly when interacting with untrusted contracts.

> For compatibility, PolkaVM maps traditional gas-related operations to its `ref_time` metric - the closest analog to Ethereum's gas system.

Reviewer: ```suggestion
For compatibility, PolkaVM maps traditional gas-related operations to its `ref_time` metric, which is the closest analog to Ethereum's gas system.
```

    The architecture establishes a constant memory limit per contract, which is the basis for calculating maximum contract nesting depth. This calculation assumes worst-case memory usage for each nested contract, resulting in a straightforward but conservative limit that operates independently of actual memory consumption. Future iterations may introduce dynamic memory metering, allowing deeper nesting depths for contracts with smaller memory footprints. However, such an enhancement would require careful implementation of cross-contract boundary limits before API stabilization, as it would introduce an additional resource metric to the system.

## Account Management - Existential Deposit

> Polkadot implements an [existential deposit](http://localhost:8000/polkadot-protocol/glossary/#existential-deposit){target=\_blank} (ED) system, requiring accounts to maintain a minimum balance to exist. When an account's balance falls below this threshold, the account is automatically deleted. This mechanism prevents state bloat from inactive accounts, unlike Ethereum, where accounts persist indefinitely regardless of balance and require no minimum funds to maintain their associated data structures (such as the nonce). This requirement extends to smart contracts, which operate as specialized accounts containing executable code.

Reviewer: ```suggestion
Polkadot implements an [existential deposit](/polkadot-protocol/glossary/#existential-deposit){target=\_blank} (ED) system, requiring accounts to maintain a minimum balance to exist. When an account's balance falls below this threshold, the account is automatically deleted. This mechanism prevents state bloat from inactive accounts, unlike Ethereum, where accounts persist indefinitely regardless of balance and require no minimum funds to maintain their associated data structures (such as the nonce). This requirement extends to smart contracts, which are specialized accounts containing executable code.
```

> Polkadot implements an [existential deposit](http://localhost:8000/polkadot-protocol/glossary/#existential-deposit){target=\_blank} (ED) system, requiring accounts to maintain a minimum balance to exist. When an account's balance falls below this threshold, the account is automatically deleted. This mechanism prevents state bloat from inactive accounts, unlike Ethereum, where accounts persist indefinitely regardless of balance and require no minimum funds to maintain their associated data structures (such as the nonce). This requirement extends to smart contracts, which operate as specialized accounts containing executable code.

Reviewer: I like the EVM approach and Polkadot approach bullet points you used in the previous section. It might be nice to do something similar here, so users can look quickly and get the details

Reviewer: Can you please check and see where else we could apply that type of formatting in the guide?

The ED system creates a scenario where each Polkadot account has an unavailable portion of its balance. This difference could cause compatibility issues with Ethereum-designed contracts and tools, mainly wallets. However, the system implements several transparent mechanisms to maintain compatibility:

Balance queries through Ethereum RPC calls automatically deduct the ED, ensuring reported balances reflect spendable amounts. Account balance checks via EVM opcodes similarly subtract the ED from reported values.

Transfers to new accounts automatically include the ED on top of the specified amount (`x + ED`). While this means the sender transfers more than the specified amount, the additional ED cost is incorporated into the transaction fee for transparency.

For contract-to-contract transfers, the system manages the ED requirement by:

> - Drawing the ED from the transaction signer rather than the sending contract

Reviewer: ```suggestion
- Drawing the ED from the transaction signer rather than the contract sending the transaction
```
